adam_core/coordinates/transform.py

```diff
@@ -1231,12 +1241,28 @@ def cartesian_to_origin(
     for origin_in in unique_origins:
 
         mask = pc.equal(coords.origin.code, origin_in).to_numpy(zero_copy_only=False)
-        vectors[mask] = get_perturber_state(
-            OriginCodes[origin_in.as_py()],
-            times[mask],
-            frame=coords.frame,
-            origin=origin,
-        ).values
+
+        # Could use try / except block here but this is more explicit
+        if origin_in.as_py() in OriginCodes.__members__:
+
+            vectors[mask] = get_perturber_state(
+                OriginCodes[origin_in.as_py()],
+                times[mask],
+                frame=coords.frame,
+                origin=origin,
+            ).values
+
+        elif origin_in.as_py() in OBSERVATORY_CODES:
+
+            vectors[mask] = get_observer_state(
+                origin_in.as_py(),
+                times[mask],
+                frame=coords.frame,
+                origin=origin,
+            ).values
+
+        else:
+            raise ValueError("Unsupported origin: {}".format(origin_in.as_py()))
```

Comment:
How many times can I repeat: `origin_in.as_py()` ... I should really just have done: `origin_str = orgin_in.as_py()` once.